src/profilers/profile-serializer.ts

```diff
@@ -15,132 +15,278 @@
  */
 
 import {perftools} from '../profile';
-import {getIndexOrAdd} from '../util';
-import {TimeProfile, TimeProfileNode} from '../v8-types';
-// A stack of function UIDs.
+import {AllocationProfileNode, ProfileNode, TimeProfile, TimeProfileNode} from '../v8-types';
+
+// A stack of function IDs.
 type Stack = Array<number>;
+
+// A function which appends the sample form of entry to samples.
+type AppendEntryToSamples =
+    (entry: Entry, samples: perftools.profiles.Sample[]) => void;
+
+// Profile node and stack trace to that node.
 interface Entry {
+  node: ProfileNode;
+  stack: Stack;
+}
+
+// Heap profile node and stack trace to that node.
+interface AllocationEntry {
+  node: AllocationProfileNode;
+  stack: Stack;
+}
+
+// Time profile node and stack trace to that node.
+interface TimeEntry {
   node: TimeProfileNode;
   stack: Stack;
 }
 
+// Fields which uniquely describe a function.
+interface FunctionKey {
+  name: string;
+  scriptId: number;
+}
+
+// Fields which uniquely describe a location.
+interface LocationKey {
+  function: FunctionKey;
+  lineNumber: number;
+  columnNumber: number;
+}
+
 /**
- * Converts v8 Profile into profile with profile format used by Stackdriver
- * Profiler.
- *
- * @param prof - profile to be converted.
- * @param intervalMicros - average time (microseconds) between samples.
+ * Used to build string table and access strings and their ids within the table
+ * when serializing a profile.
  */
-export function serializeTimeProfile(
-    prof: TimeProfile, intervalMicros: number) {
-  let samples: Array<perftools.profiles.Sample> = [];
-  let locations: Array<perftools.profiles.Location> = [];
-  let functions: Array<perftools.profiles.Function> = [];
-  let locationMap: Map<number, perftools.profiles.Location> = new Map();
-  let functionMap: Map<number, perftools.profiles.Function> = new Map();
-  let strings = [''];
+class StringTable {
+  strings: string[];
+  stringsMap: Map<string, number>;
 
-  let sampleValueType = createSampleValueType();
-  let timeValueType = createTimeValueType();
+  constructor() {
+    this.strings = [];
+    this.stringsMap = new Map<string, number>();
+  }
 
-  serializeNode(prof.topDownRoot);
+  /**
+   * Returns index of str within the table. Also adds str to string table if
+   * str is not in the table already.
+   *
+   * @param str
```

Comment:
Is there any value in having this empty parameter annotation? Is this the linter requirement or something?